File: apps/backend/app/schemas/compliance.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


# ── Literals ─────────────────────────────────────────────────────────────────

CheckStatusType  = Literal["pending", "running", "passed", "flagged", "blocked", "error"]
CheckModeType    = Literal["rule", "ai", "both"]
RiskCategoryType = Literal["ad_safety", "copyright_risk", "factual_risk", "reused_content", "ai_disclosure"]
RiskSeverityType = Literal["critical", "high", "medium", "low", "info"]
FlagSourceType   = Literal["rule", "ai"]


# ── RiskFlag ──────────────────────────────────────────────────────────────────

class RiskFlagRead(BaseModel):
    id: uuid.UUID
    check_id: uuid.UUID
    category: RiskCategoryType
    severity: RiskSeverityType
    source: FlagSourceType
    rule_id: str
    title: str
    detail: str
    evidence: str | None
    suggestion: str | None
    text_start: int | None
    text_end: int | None
    is_dismissed: bool
    dismissed_by: str | None
    dismissed_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class RiskFlagDismiss(BaseModel):
    dismissed_by: str = Field(min_length=1, max_length=255)
    reason: str | None = None


# ── ComplianceCheck ───────────────────────────────────────────────────────────

class ComplianceCheckRead(BaseModel):
    id: uuid.UUID
    channel_id: uuid.UUID
    script_id: uuid.UUID | None
    publication_id: uuid.UUID | None
    mode: CheckModeType
    status: CheckStatusType
    risk_score: float
    category_scores: dict[str, float]
    flag_count: int
    critical_count: int
    high_count: int
    monetization_eligible: bool
    ai_disclosure_required: bool
    is_overridden: bool
    override_by: str | None
    override_reason: str | None
    overridden_at: datetime | None
    started_at: datetime | None
    completed_at: datetime | None
    flags: list[RiskFlagRead] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ComplianceCheckCreate(BaseModel):
    script_id: uuid.UUID | None = None
    publication_id: uuid.UUID | None = None
    mode: CheckModeType = "both"


class ComplianceCheckOverride(BaseModel):
    override_by: str = Field(min_length=1, max_length=255)
    override_reason: str = Field(min_length=10, max_length=2000)


# ── Summary views ─────────────────────────────────────────────────────────────

class CategoryBreakdown(BaseModel):
    category: RiskCategoryType
    score: float
    flag_count: int
    worst_severity: RiskSeverityType | None
    flags: list[RiskFlagRead]


class ComplianceCheckDetail(ComplianceCheckRead):
    """Full check result with per-category breakdown."""
    categories: list[CategoryBreakdown]


class ComplianceSummary(BaseModel):
    """Lightweight summary for list views."""
    id: uuid.UUID
    script_id: uuid.UUID | None
    publication_id: uuid.UUID | None
    status: CheckStatusType
    risk_score: float
    flag_count: int
    critical_count: int
    monetization_eligible: bool
    ai_disclosure_required: bool
    is_overridden: bool
    created_at: datetime

File: apps/frontend/src/lib/api/compliance.ts
import type {
  ComplianceCheck,
  ComplianceCheckDetail,
  ComplianceSummary,
  RiskFlag,
} from "@/lib/types";
import { apiClient } from "./client";

export const complianceApi = {
  runCheck: (
    channelId: string,
    data: { script_id?: string; publication_id?: string; mode?: string }
  ) =>
    apiClient.post<ComplianceCheck>(
      `/compliance/channels/${channelId}/checks`,
      data
    ),

  getCheck: (checkId: string) =>
    apiClient.get<ComplianceCheckDetail>(`/compliance/checks/${checkId}`),

  listChecks: (channelId: string, params?: { script_id?: string; status?: string; limit?: number }) => {
    const q = new URLSearchParams();
    if (params?.script_id) q.set("script_id", params.script_id);
    if (params?.status)    q.set("status", params.status);
    if (params?.limit)     q.set("limit", String(params.limit));
    return apiClient.get<ComplianceSummary[]>(
      `/compliance/channels/${channelId}/checks${q.toString() ? `?${q}` : ""}`
    );
  },

  latestForScript: (scriptId: string) =>
    apiClient.get<ComplianceCheck | null>(`/compliance/scripts/${scriptId}/latest-check`),

  override: (checkId: string, data: { override_by: string; override_reason: string }) =>
    apiClient.post<ComplianceCheck>(`/compliance/checks/${checkId}/override`, data),

  dismissFlag: (flagId: string, data: { dismissed_by: string; reason?: string }) =>
    apiClient.post<RiskFlag>(`/compliance/flags/${flagId}/dismiss`, data),
};
